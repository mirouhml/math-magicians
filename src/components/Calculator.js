/* eslint-disable max-len */

import React from 'react';
import calculate from '../logic/calculate';
import Button from './Button';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.buttons = [
      ['AC', '+/-', '%', 'รท'],
      ['7', '8', '9', 'ร'],
      ['4', '5', '6', '-'],
      ['1', '2', '3', '+'],
      ['0', '.', '='],
    ];
    this.state = {};
  }

  componentDidMount() {
    this.setState = ({
      total: 0,
      next: null,
      operation: null,
    });
  }

  onClick(button) {
    const state = calculate(this.state, button);
    this.setState = state;
  }

  render() {
    return (
      <table>
        <tr>
          <td className="result" colSpan="4">0</td>
        </tr>
        <Button buttons={this.buttons[0]} action={this.onClick} last={false} />
        <Button buttons={this.buttons[1]} action={this.onClick} last={false} />
        <Button buttons={this.buttons[2]} action={this.onClick} last={false} />
        <Button buttons={this.buttons[3]} action={this.onClick} last={false} />
        <Button buttons={this.buttons[4]} action={this.onClick} last />
      </table>
    );
  }
}

export default Calculator;
