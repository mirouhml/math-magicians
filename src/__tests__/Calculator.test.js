import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

describe('Calculator component tests:', () => {
  it('Calculator component renders correctly', () => {
    const tree = renderer
      .create(<Calculator />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Perform an operation and display its result - test addition 16+23 = 39', () => {
    const { container } = render(<Calculator />);
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('6'));
    fireEvent.click(screen.getByText('+'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('='));
    const result = container.querySelector('.result').textContent;
    expect(result).toBe('39');
  });

  it('Perform an operation and display its result - test substraction -4-16 = -20', () => {
    const { container } = render(<Calculator />);
    fireEvent.click(screen.getByText('4'));
    fireEvent.click(screen.getByText('+/-'));
    fireEvent.click(screen.getByText('-'));
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('6'));
    fireEvent.click(screen.getByText('='));
    const result = container.querySelector('.result').textContent;
    expect(result).toBe('-20');
  });

  it('Perform an operation and display its result - test multiplication 9x10 = 90', () => {
    const { container } = render(<Calculator />);
    fireEvent.click(screen.getByText('9'));
    fireEvent.click(screen.getByText('ร'));
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('='));
    const result = container.querySelector('.result').textContent;
    expect(result).toBe('90');
  });

  it('Perform an operation and display its result - test devision 100รท2 = 50', () => {
    const { container } = render(<Calculator />);
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('รท'));
    fireEvent.click(screen.getByText('2'));
    fireEvent.click(screen.getByText('='));
    const result = container.querySelector('.result').textContent;
    expect(result).toBe('50');
  });

  it('Perform an operation and display its result - test devision by zero', () => {
    const { container } = render(<Calculator />);
    fireEvent.click(screen.getByText('1'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('รท'));
    fireEvent.click(screen.getByText('0'));
    fireEvent.click(screen.getByText('='));
    const result = container.querySelector('.result').textContent;
    expect(result).toBe('Can\'t divide by 0.');
  });
});
