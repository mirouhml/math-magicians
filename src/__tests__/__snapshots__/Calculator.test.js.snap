// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component tests: Calculator component renders correctly 1`] = `
<table
  className="calculator-table"
>
  <tbody>
    <tr>
      <td
        className="result"
        colSpan="4"
      >
        0
      </td>
    </tr>
    <tr>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          %
        </button>
      </td>
      <td>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          รท
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          7
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          8
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          9
        </button>
      </td>
      <td>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          ร
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          5
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          6
        </button>
      </td>
      <td>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </td>
    </tr>
    <tr>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </td>
      <td>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </td>
    </tr>
    <tr>
      <td
        colSpan="2"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          0
        </button>
      </td>
      <td>
        <button
          onClick={[Function]}
          type="button"
        >
          .
        </button>
      </td>
      <td>
        <button
          className="operator"
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
